pkg/binlog/event/util.go

```diff
@@ -269,3 +271,170 @@ func writeStringColumnValue(buf *bytes.Buffer, value interface{}) error {
 	}
 	return terror.ErrBinlogWriteBinaryData.Delegate(err)
 }
+
+// https://dev.mysql.com/doc/internals/en/query-event.html
+const (
+	QFlags2Code = iota
+	QSqlModeCode
+	QCatalog
+	QAutoIncrement
+	QCharsetCode
+	QTimeZoneCode
+	QCatalogNzCode
+	QLcTimeNamesCode
+	QCharsetDatabaseCode
+	QTableMapForUpdateCode
+	QMasterDataWrittenCode
+	QInvokers
+	QUpdatedDbNames
+	QMicroseconds
+)
+
+var (
+	// https://dev.mysql.com/doc/internals/en/query-event.html
+	statusVarsFixedLength = map[byte]int{
+		QFlags2Code:            4,
+		QSqlModeCode:           8,
+		QAutoIncrement:         2 + 2,
+		QCharsetCode:           2 + 2 + 2,
+		QLcTimeNamesCode:       2,
+		QCharsetDatabaseCode:   2,
+		QTableMapForUpdateCode: 8,
+		QMasterDataWrittenCode: 4,
+		QMicroseconds:          3,
+	}
+)
+
+// GetAnsiQuotesMode gets ansi-quotes mode from binlog statusVars
+func GetAnsiQuotesMode(statusVars []byte) (bool, error) {
+	vars, err := statusVarsToKV(statusVars)
+	if err != nil {
+		return false, err
+	}
+	b, ok := vars[QSqlModeCode]
+	if !ok {
+		// TODO(lance6716): if this will happen, create a terror
+		return false, errors.New("Q_SQL_MODE_CODE not found in status_vars")
+	}
+
+	r := bytes.NewReader(b)
+	var v int64
+	_ = binary.Read(r, binary.LittleEndian, &v)
+
+	// MODE_ANSI_QUOTES = 0x00000004 ref: https://dev.mysql.com/doc/internals/en/query-event.html#q-sql-mode-code
+	return v&0x00000004 != 0, nil
+}
+
+func statusVarsToKV(statusVars []byte) (map[byte][]byte, error) {
+	r := bytes.NewReader(statusVars)
+	vars := make(map[byte][]byte)
+	var (
+		value []byte
+	)
+
+	// NOTE: this closure modifies variable `value`
+	appendLengthThenCharsToValue := func() error {
+		length, err := r.ReadByte()
+		if err != nil {
+			return err
+		}
+		value = append(value, length)
+
+		buf := make([]byte, length)
+		n, err := r.Read(buf)
+		if err != nil {
+			return err
+		}
+		if n != int(length) {
+			return errors.New("unexpected EOF")
+		}
+		value = append(value, buf...)
+		return nil
+	}
+
+	generateError := func() (map[byte][]byte, error) {
+		offset, _ := r.Seek(0, io.SeekCurrent)
+		return nil, terror.ErrBinlogStatusVarsParse.Generate(statusVars, offset)
+	}
+
+	for {
+		// reset value
+		value = make([]byte, 0)
+		key, err := r.ReadByte()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return generateError()
+		}
+
+		if _, ok := vars[key]; ok {
+			return generateError()
+		}
+
+		if length, ok := statusVarsFixedLength[key]; ok {
+			value = make([]byte, length)
+			n, err := r.Read(value)
+			if err != nil || n != length {
+				return generateError()
+			}
+
+			vars[key] = value
+			continue
+		}
+
+		// get variable-length value of according key and save it in `value`
+		switch key {
+		// 1-byte length + <length> chars of the catalog + '0'-char
+		case QCatalog:
+			if err = appendLengthThenCharsToValue(); err != nil {
+				return generateError()
```

Comment:
can we carry `err` in the returned error?

Reply:
changed in https://github.com/pingcap/dm/pull/882/commits/7178f200203c5a86b14a2ea8009a6543bf27e990